### PR TITLE
Add pedestal transport model section to CombinedTransportModel

### PR DESCRIPTION
**Pull request by theo-brown**
- [x] Add separate pedestal model to CombinedTransportModel
- [x] Update tests
- [x] Update docs 
- [x] Add test to check behaviour when no pedestal model supplied - intended behaviour is revert to chi_min, etc 

Implements #941

### CODE REVIEW

File: torax/_src/transport_model/combined.py [hunk 1]
```diff
@@ -33,19 +35,72 @@
...
+    # Apply min/max clipping
+    transport_coeffs = self._apply_clipping(
+        transport_runtime_params,
+        transport_coeffs,
+    )
+
+    # In contrast to the base TransportModel, we do not apply patches
+
+    # Return smoothed coefficients if smoothing is enabled
+    # TODO: what should be done about masking? In base TransportModel, the
```
Comment (theo-brown) [inline torax/_src/transport_model/combined.py]: What should be done about masking? Need to compare this method with what was there before.

Comment (theo-brown) [reply torax/_src/transport_model/combined.py]: Resolved, see updated comment

File: torax/_src/transport_model/pydantic_model.py [hunk 2]
```diff
@@ -368,50 +368,69 @@ class CombinedTransportModel(pydantic_model_base.TransportBase):
 
   Attributes:
     model_name: The transport model to use. Hardcoded to 'combined'.
-    transport_models: A sequence of transport models, whose outputs will be
-      summed to give the combined transport coefficients.
+    core_transport_models: A sequence of transport models, whose outputs will be
+      summed to give the combined core transport coefficients.
```
Comment (tamaranorman) [inline torax/_src/transport_model/pydantic_model.py:372]: This change is technically breaking for anyone with a config that uses a CombinedTransportModel and technically we have agreed not to change that. This was only recently added but we should decide if we think we want to allow this.

File: torax/_src/transport_model/pydantic_model.py [hunk 3]
```diff
@@ -368,50 +368,69 @@ class CombinedTransportModel(pydantic_model_base.TransportBase):
...
+    pedestal_transport_model: A model for pedestal transport coefficients.
   """
 
-  transport_models: Sequence[CombinedCompatibleTransportModel] = pydantic.Field(
+  core_transport_models: Sequence[
+      CombinedCompatibleTransportModel
+  ] = pydantic.Field(
       default_factory=list
   )  # pytype: disable=invalid-annotation
+  pedestal_transport_model: CombinedCompatibleTransportModel
```
Comment (tamaranorman) [inline torax/_src/transport_model/pydantic_model.py]: Is this required? Or should we still support the old post pedestal model if none is provided using the coefficients? We have a default for the core_transport_models which seems to be an empty list? What is the point of that, or should they both be required

File: torax/_src/transport_model/pydantic_model.py [hunk 4]
```diff
@@ -368,50 +368,69 @@ class CombinedTransportModel(pydantic_model_base.TransportBase):
...
+
     return combined.DynamicRuntimeParams(
-        transport_model_params=model_params_list,
+        core_transport_model_params=core_transport_model_params,
+        pedestal_transport_model_params=pedestal_transport_model_params,
         **base_kwargs,
     )
 
   @pydantic.model_validator(mode='after')
   def _check_fields(self) -> typing_extensions.Self:
```
Comment (tamaranorman) [inline torax/_src/transport_model/pydantic_model.py:421]: Do we need some validation on the pedestal model inputs

File: docs/scripts/combined_transport_example.py [hunk 5]
```diff
@@ -51,13 +51,13 @@ def main(argv: Sequence[str]) -> None:
...
-                  'chi_i': 0.5,
-                  'rho_min': 0.5,
-                  'rho_max': 1.0,
-              },
           ],
+          'pedestal_transport_model': {
+              'model_name': 'constant',
+              'chi_i': 0.5,
+              'rho_min': 0.5,
+              'rho_max': 1.0,
```
Comment (tamaranorman) [inline docs/scripts/combined_transport_example.py]: Should rho be set or should we just have it be everything after the pedestal?

File: docs/scripts/combined_transport_example.py [hunk 6]
```diff
@@ -51,13 +51,13 @@ def main(argv: Sequence[str]) -> None:
                   'rho_min': 0.2,
                   'rho_max': 0.5,
```
Comment (tamaranorman) [inline docs/scripts/combined_transport_example.py]: Do we need errors for regions not covered by a transport model?

File: docs/scripts/combined_transport_example.py
Comment (tamaranorman) [inline docs/scripts/combined_transport_example.py:1]: Was this not used to generate something in the docs, does that now need updating

File: torax/_src/transport_model/combined.py [hunk 7]
```diff
@@ -100,20 +155,63 @@ def _call_implementation(
...
 
   def __eq__(self, other):
     return (
         isinstance(other, CombinedTransportModel)
-        and self.transport_models == other.transport_models
+        and self.core_transport_models == other.core_transport_models
+        and self.pedestal_transport_model == other.pedestal_transport_model
+    )
+
+  def _apply_pedestal_domain_restriction(
```
Comment (tamaranorman) [inline torax/_src/transport_model/combined.py:181]: nit: Move logic above the hash and equals models to keep them altogether

File: docs/scripts/combined_transport_example.py (hunk 6 above)
Comment (jcitrin) [reply docs/scripts/combined_transport_example.py]: It's a valid option in the sense of it's "OK" for a zone to have no turbulent transport defined. That zone will still have set the minimum transport coefficient values, and in the PDE itself (in calc_coeffs.py) there may be the addition of neoclassical transport (if on). While there's only limited use-cases of having no turbulent transport in a specific zone (maybe internal-transport-barrier explorations?), I don't think we need errors if this occurs

Comment (jcitrin) [reply docs/scripts/combined_transport_example.py]: That being said I'm not sure that this is a good example. rho_max should be 1.0, and the pedestal model will supersede it in the pedestal region. See comment below

File: docs/scripts/combined_transport_example.py (hunk 5 above)
Comment (jcitrin) [reply docs/scripts/combined_transport_example.py]: Agree. A pedestal transport model should be in the pedestal zone as set by the 'pedestal' config. We shouldn't have two separate definitions of where the pedestal lives. pedestal_transport_models thus should not have a rho_min and rho_max.

File: torax/_src/transport_model/combined.py [hunk 8]
```diff
@@ -100,20 +155,63 @@ def _call_implementation(
...
+        pedestal_transport_coeffs,
+        pedestal_model_output,
+    )
 
+    # Combine the transport coefficients from core and pedestal models.
     combined_transport_coeffs = jax.tree.map(
         lambda *leaves: sum(leaves),
-        *component_transport_coeffs_list,
+        *(core_transport_coeffs_list + [pedestal_transport_coeffs]),
     )
```
Comment (jcitrin) [inline torax/_src/transport_model/combined.py]: There should be symmetry with core_transport_coeffs in the sense that multiple pedestal_transport_models should also be allowed

File: torax/_src/transport_model/combined.py (hunk 7 above)
Comment (jcitrin) [inline torax/_src/transport_model/combined.py:181]: Doesn't need to be for this PR, but it feels wasteful to run a pedestal model for the entire radius, when in fact most of the indices will be set 0, since the pedestal region is small.

We may want to refactor transport models to only do calculations within the domain of interest. Likely this can also be made vmappable etc. Can add a TODO to "combined"

File: torax/_src/transport_model/pydantic_model.py (hunk 2 above)
Comment (jcitrin) [reply torax/_src/transport_model/pydantic_model.py:372]: agree. Should revert this. It's ok to have "transport_models" and "pedestal_transport_models". In V2 we will remove combined altogether by essentially making it default, and then having "core_transport_models" and "pedestal_transport_models"

File: torax/_src/transport_model/pydantic_model.py (hunk 3 above)
Comment (jcitrin) [reply torax/_src/transport_model/pydantic_model.py]: Agree. If None is provided for pedestal_transport_model, should revert to the De_min, chi_min default.

**Review by jcitrin — COMMENTED** [review]
In addition to the comments, please update the docs following the converged-upon patterns.

File: torax/_src/transport_model/pydantic_model.py (hunk 2 above)
Comment (theo-brown) [reply torax/_src/transport_model/pydantic_model.py:372]: Sure - I had forgotten that Combined had already made its way into a minor patch. Will revert and add a note about v2

File: torax/_src/transport_model/pydantic_model.py [hunk 9]
```diff
@@ -368,50 +368,69 @@ class CombinedTransportModel(pydantic_model_base.TransportBase):
...
     super()._check_fields()
-    if not self.transport_models:
+    if not self.core_transport_models:
       raise ValueError(
-          'transport_models cannot be empty for CombinedTransportModel. '
+          'core_transport_models cannot be empty for CombinedTransportModel. '
           'Please provide at least one transport model configuration.'
       )
```
Comment (theo-brown) [inline torax/_src/transport_model/pydantic_model.py]: As per Jonathan's comment, there is a conceivable scenario where only neoclassical transport is used and no turbulent transport. Hence, removing this as an unnecessary check.

File: docs/scripts/combined_transport_example.py
Comment (theo-brown) [reply docs/scripts/combined_transport_example.py:1]: Yep, but was just leaving docs changes until source code changes had finalised!

File: torax/_src/transport_model/combined.py (hunk 7 above)
Comment (theo-brown) [reply torax/_src/transport_model/combined.py:181]: Agreed, this was a very botched way of doing it. My reasoning was trying to avoid creating a new base class for PedestalTransportModels and so it made most sense to me to just replicate what was done in TransportModel. I'll add a comment that this should be replaced with something more efficient.

Comment (theo-brown) [reply torax/_src/transport_model/combined.py:181]: On the topic of vmappability - is switching all TransportModels to be vmapped over their domain going to cause problems with different surrogates people may build? My concern is to do with management of state and stuff, like whether to split RNGs or not. It makes total sense for the deterministic physics calculation models (e.g. BgB), but could it be a potential footgun for surrogates?

Could be trivial to navigate but would have to be documented as a requirement for models _call_implementation.

Comment (theo-brown) [reply torax/_src/transport_model/combined.py:181]: Also, are you suggesting vmapping over the domain or over the models? Not sure we could vmap over heterogeneous structures (ie, different models)

Comment (jcitrin) [reply torax/_src/transport_model/combined.py:181]: Not sure about stochastic surrogates. Of course surrogates can also be deterministic (and the ones we use currently are). It's a subclass of surrogates that would include stochastic elements. Instinct is "fix it when it's a problem", there are likely solutions.

On the mapped index, most natural one for now corresponds to the radial coordinate.

File: torax/_src/transport_model/pydantic_model.py (hunk 3 above)
Comment (theo-brown) [reply torax/_src/transport_model/pydantic_model.py]: Added a test to check for this behaviour.

File: torax/_src/transport_model/combined.py (hunk 7 above)
Comment (theo-brown) [reply torax/_src/transport_model/combined.py:181]: I've done a small refactor to improve code reuse. I think your suggested change is good but it should be in another PR? Made an issue, #1347

File: torax/_src/transport_model/pydantic_model.py (hunk 4 above)
Comment (theo-brown) [reply torax/_src/transport_model/pydantic_model.py:421]: Added checks that patches, rho_min, and rho_max are not set on pedestal models

**Review by tamaranorman — APPROVED** [review]

File: docs/configuration.rst [hunk 10]
```diff
@@ -1127,8 +1141,8 @@ This would produce a ``chi_i`` profile that looks like the following.
 Note that in the region :math:`[0, 0.2]`, only the first component is active,
 so ``chi_i = 1.0``. In :math:`(0.2, 0.3]` the first two components are both
 active, leading to a combined value of ``chi_i = 3.0``. In :math:`(0.3, 0.5]`,
-only the second model is active (``chi_i = 2.0``), and in :math:`(0.5, 1.0]`
-only the fourth model is active (``chi_i = 0.5``).
+only the second model is active (``chi_i = 2.0``), and in :math:`(0.8, 1.0]`
+only the pedestal transport model is active (``chi_i = 0.5``).
```
Comment (jcitrin) [inline docs/configuration.rst]: This is incorrect. From rho 0.3 to the default rho_norm_ped_top of 0.91, only the second transport_model is active. Then, from the default rho_norm_ped_top=0.91 to the edge, the pedestal model is active with chi_i=0.5.

Recommend explicitly setting `rho_norm_ped_top` in 'pedestal' to some value like 0.9 to make the zones more explicitl

File: docs/scripts/combined_transport_example.py [hunk 11]
```diff
@@ -33,7 +35,12 @@ def main(argv: Sequence[str]) -> None:
           'geometry_type': 'circular',
           'n_rho': 30,  # for higher resolution plotting
       },
-      'pedestal': {},
+      'pedestal': {
```
Comment (jcitrin) [inline docs/scripts/combined_transport_example.py:37]: Recommend setting `rho_norm_ped_top` here

**Review by jcitrin — APPROVED** [review]
Looks good! Following the small fixes + fixing the merge conflicts, will bring it it

File: docs/configuration.rst (hunk 10 above)
Comment (theo-brown) [reply docs/configuration.rst]: Good catch, not sure why I made this mistake! Added `rho_norm_ped_top` and changed the explanation to fix. I regenerated the plot but it still looks the same, presumably because the profile is on the discrete rho grid and thus can't resolve the difference between ρ=0.91 and ρ=0.9.

**Comment by theo-brown** [comment]
I've made the changes, rebased onto main, and squashed into a single commit, let me know if there's anything more that needs doing!

**Review by jcitrin — APPROVED** [review]
